src/include/migraphx/heap.hpp

```diff
@@ -0,0 +1,49 @@
+#ifndef MIGRAPHX_GUARD_RTGLIB_HEAP_HPP
+#define MIGRAPHX_GUARD_RTGLIB_HEAP_HPP
+
+#include <thread>
+#include <cmath>
+#include <algorithm>
+#include <vector>
+#include <cassert>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+template <class T, class Compare>
+struct heap
+{
+    std::vector<T> data;
+    Compare compare;
+
+    heap(const std::vector<T>& val, Compare comp) : data(val), compare(std::move(comp))
+    {
+        for(int i = 2; i <= data.size(); ++i)
+        {
+            std::push_heap(data.begin(), data.begin() + i, compare);
+        }
+    }
+
+    void update(const T val)
+    {
+        std::pop_heap(data.begin(), data.end(), compare);
+        if(compare(val, data.back()))
+            data.back() = val;
+        std::push_heap(data.begin(), data.end(), compare);
+    }
+
+    void sort() { std::sort_heap(data.begin(), data.end(), compare); }
+
+    std::vector<T>& get() { return data; }
+};
+
+template <class T, class Compare>
+heap<T, Compare> make_heap(std::vector<T>& val, Compare compare)
+{
+    return {std::move(val), std::move(compare)};
+}
```

Comment:
Can we just put this in the topk header as this is specific to opk? Even better if this can be nested under topk.

Reply:
These APIs are the same as that provided by the STL library. I think it is a general heap implementation.

Reply:
Yea but the `update` method is very specific to topk(as we only push the value onto the heap if it will be a top value).

Reply:
Also, `priority_queue` provides a standard API for a heap, but it doesn't provide `sort` or `get` methods.